ocaml/typing/persistent_env.ml

```diff
@@ -344,23 +397,280 @@ let find_import ~allow_hidden penv ~check modname =
           add_import penv modname;
           acknowledge_import penv ~check modname psig
 
+let remember_global { globals; _ } global ~mentioned_by =
+  if Global_module.has_arguments global then
+    let global_name = Global_module.to_name global in
+    match Hashtbl.find globals global_name with
+    | exception Not_found ->
+        Hashtbl.add globals global_name
+          { gn_global = global; gn_mentioned_by = mentioned_by }
+    | { gn_global = old_global; gn_mentioned_by = first_mentioned_by } ->
+        if not (Global_module.equal old_global global) then
+          error (Inconsistent_global_name_resolution {
+              name = global_name;
+              old_global;
+              new_global = global;
+              first_mentioned_by;
+              now_mentioned_by = mentioned_by;
+            })
+
+let current_unit_is0 name ~allow_args =
+  match CU.get_current () with
+  | None -> false
+  | Some current ->
+      match CU.to_global_name current with
+      | Some { head; args } ->
+          (args = [] || allow_args)
+          && CU.Name.equal name (head |> CU.Name.of_string)
+      | None -> false
+
+let current_unit_is name =
+  current_unit_is0 name ~allow_args:false
+
+let current_unit_is_instance_of name =
+  current_unit_is0 name ~allow_args:true
+
 (* Enforce the subset rule: we can only refer to a module if that module's
    parameters are also our parameters. *)
-let check_for_unset_parameters penv modname import =
+let check_for_unset_parameters penv global =
+  (* A hidden argument specifies that the importing module should forward a
+     parameter to the imported module. Therefore it's the hidden arguments that
+     we need to check. *)
   List.iter
-    (fun param ->
-       if not (is_registered_parameter_import penv param) then
+    (fun ({ param = arg_name; value = arg_value } : Global_module.argument) ->
+       (* The _value_ is what we care about - the name lives in the imported
+          module's namespace, not ours *)
+       ignore arg_name;
+       let value_name = Global_module.to_name arg_value in
+       if not (is_registered_parameter_import penv value_name) then
          error (Imported_module_has_unset_parameter {
-             imported = modname;
-             parameter = param;
+             imported = Global_module.to_name global;
+             parameter = value_name;
            }))
-    import.imp_params
+    global.Global_module.hidden_args;
+  (* The visible arguments in [global] must also satisfy the subset rule.
+     However, these will already have been checked since [compute_global] loads
+     all argument names and values. *)
+  ()
```

Comment:
Yeah, what I'm going for here is “we're actually done, but here's a comment about why.” Maybe I should just add the sentence “Therefore there's nothing more to do” to the comment so that the `()` more obviously means “do nothing”?

Reply:
Adding the sentence.